addon/-private/system/schema.js

```diff
@@ -0,0 +1,295 @@
+    let options = propertyMeta.options;
+    if (options.inverse === null) {
+      return null;
+    }
+
+    let inverseName;
+    let inverseKind;
+    let inverse;
+
+    // If inverse is specified manually, return the inverse
+    if (options.inverse) {
+      inverseName = options.inverse;
+      inverse = inverseSchema.relationships[inverseName];
+
+      assert("We found no inverse relationships by the name of '" + inverseName + "' on the '" + inverseSchema.modelName +
+        "' model. This is most likely due to a missing attribute on your model definition.", !Ember.isNone(inverse));
+
+      inverseKind = inverse.kind;
+    } else {
+      //No inverse was specified manually, we need to use a heuristic to guess one
+      if (propertyMeta.type === propertyMeta.parentType.modelName) {
+        warn(`Detected a reflexive relationship by the name of '${name}' without an inverse option. Look at http://emberjs.com/guides/models/defining-models/#toc_reflexive-relation for how to explicitly specify inverses.`, false, {
+          id: 'ds.model.reflexive-relationship-without-inverse'
+        });
+      }
+
+      let possibleRelationships = findPossibleInverses(this, inverseSchema);
+
+      if (possibleRelationships.length === 0) {
+        return null;
+      }
+
+      let filteredRelationships = possibleRelationships.filter((possibleRelationship) => {
+        let optionsForRelationship = inverseSchema.metaForProperty(possibleRelationship.name).options;
+
+        return name === optionsForRelationship.inverse;
+      });
+
+      assert("You defined the '" + name + "' relationship on model:" + this.modelName + ", but you defined the inverse relationships of model:" +
+        inverseSchema.modelName + " multiple times. Look at http://emberjs.com/guides/models/defining-models/#toc_explicit-inverses for how to explicitly specify inverses",
+        filteredRelationships.length < 2);
+
+      if (filteredRelationships.length === 1 ) {
+        possibleRelationships = filteredRelationships;
+      }
+
+      assert("You defined the '" + name + "' relationship on model:" + this.modelName + ", but multiple possible inverse relationships of model:" +
+        this.modelName + " were found on model:" + inverseSchema.modelName + ". Look at http://emberjs.com/guides/models/defining-models/#toc_explicit-inverses for how to explicitly specify inverses",
+        possibleRelationships.length === 1);
+
+      inverseName = possibleRelationships[0].name;
+      inverseKind = possibleRelationships[0].kind;
+    }
+
+    return {
+      type: inverseSchema,
+      name: inverseName,
+      kind: inverseKind
+    };
+  }
+
+  metaForProperty(propertyName) {
+    return this.properties[propertyName];
+  }
+
+  get recordMap() {
+    if (this._recordMap === null) {
+      this._recordMap = {
+        idToRecord: new EmptyObject(),
+        records: [],
+        metadata: new EmptyObject(),
+        type: this.model
```

Comment:
Are you keeping `type` here and bellow because it would be too much work for now to refactor everything?